shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java

```diff
@@ -188,39 +269,65 @@ private void invokeHandler(
     }
   }
 
-  @Override
-  public void handleMessageFromDart(
-      @NonNull final String channel,
+  private void dispatchMessageToQueue(
+      @NonNull String channel,
+      @Nullable HandlerInfo handlerInfo,
       @Nullable ByteBuffer message,
-      final int replyId,
+      int replyId,
       long messageData) {
-    // Called from the ui thread.
-    Log.v(TAG, "Received message from Dart over channel '" + channel + "'");
-    @Nullable final HandlerInfo handlerInfo = messageHandlers.get(channel);
-    @Nullable
     final DartMessengerTaskQueue taskQueue = (handlerInfo != null) ? handlerInfo.taskQueue : null;
     Runnable myRunnable =
         () -> {
           Trace.beginSection("DartMessenger#handleMessageFromDart on " + channel);
           try {
             invokeHandler(handlerInfo, message, replyId);
             if (message != null && message.isDirect()) {
-              // This ensures that if a user retains an instance to the ByteBuffer and it happens to
-              // be direct they will get a deterministic error.
+              // This ensures that if a user retains an instance to the ByteBuffer and it
+              // happens to be direct they will get a deterministic error.
               message.limit(0);
             }
           } finally {
+            Trace.endSection();
             // This is deleting the data underneath the message object.
             flutterJNI.cleanupMessageData(messageData);
-            Trace.endSection();
           }
         };
-    @NonNull
     final DartMessengerTaskQueue nonnullTaskQueue =
         taskQueue == null ? platformTaskQueue : taskQueue;
     nonnullTaskQueue.dispatch(myRunnable);
   }
 
+  @Override
+  public void handleMessageFromDart(
+      @NonNull String channel, @Nullable ByteBuffer message, int replyId, long messageData) {
+    // Called from the ui thread.
+    Log.v(TAG, "Received message from Dart over channel '" + channel + "'");
+
+    HandlerInfo handlerInfo;
+    synchronized (handlersLock) {
+      if (messageHandlers.contains(channel)) {
```

Comment:
If the message was not deferred, then it must be dispatched even if the handlerInfo is null.

Try something like:
```
synchronized (handlersLock) {
  handlerInfo = messageHandlers.get(channel)
  messageDeferred = (canDelayTasks && handlerInfo == null);
  if (messageDeferred) {
    // add it to the deferred message list
  }
}
if (!messageDeferred) {
  // process the message now
}

Reply:
done